proposals/2024-05-23_controller_engine_v2.md

```diff
@@ -0,0 +1,256 @@
+    the same mapping is important for a good user experience and to allow for
+    more complex mappings.
+ 2. The current controller engine (`ControllerScriptEngineLegacy`) is not able
+    to support modules of any kind. This is important for code organization and
+    reusability. In order to increase mapping code quality and maintainability,
+    it is important to be able to split the mapping code into multiple files and
+    to be able to reuse code between different mappings. This is especially
+    important now that mapping authors are starting to reuse functionality
+    across different hardware devices (see Numark NS6II and Mixtrack variants or
+    similarity between different pioneer controllers)
+ 3. Current mappings modify global state and assume that they are the only
+    mapping running. This needs to be fixed in order to support multiple
+    mappings running at the same time in the same "execution engine".
+ 4. Device hotplug is currently not possible / hard to implement. This is
+    primarily because of the underlying protocol IO layer not willing to support
+    it. This document proposes an alternative approach that lets us more easily
+    switch the underlying library.
+ 5. Allow changing the manifest format. Many people have expressed distaste in
+    the XML manifest format. Careful implementation of the new format should
+    allow experimentation with different formats.
+ 6. Changing out infrastructure like this is virtually impossible. The proposed
+    architecture is designed to be modular and allows for all the required
+    components to be implemented gradually.
+
+## Goals
+
+Goals and use cases for the solution as proposed in [How](#how):
+
+* Allow multiple controllers to be used at the same time.
+* Mappings should be able to organize themselves into modules.
+* IO protocols should easily be changeable and not not be tied to the mapping.
+* Allow easy, gradual implementation of the system.
+
+### Audience
+
+This is primarily targeted at developers and power users who are interested in
+the controller engine. End-users only interested in the point-and-click mapping
+editor should not experience any significant changes.
+
+## Non-Goals
+
+* Replace `ControllerScriptEngine` in the short term until we are confident in
+  the new design.
+* make changes to the scratching code as it is orthogonal to this proposal.
+* Make changes to the effects system as it is orthogonal to this proposal.
+
+## How
+
+Explain the full overview of the proposed solution. Some guidelines:
+
+### File Structure
+
+Built-in mappings reside in `res/controllers/` where each mapping is a directory
+containing at least a manifest file named `manifest.xxx` (where `.xxx` is the
+language-specific file extension (eg. `.xml`, `.yml`, `.json`, etc)).
+Additionally, `res/controllers/lib` contains shared modules that can be used by
+multiple mappings (such as componentsJS).
```

Comment:
We can even allow to share mappings via an unpacked zip file with a fancy extension that Mixxx is able to unpack. 

I just remember discussions that mapping writers want to work only in a single file. So we need to be clear about our goals here.

Reply:
I personally find it a little silly to zip files and then use a custom extension (looks like a bad attempt at hiding something) but from a technical perspective, I don't see anything wrong with that. Its just an extra step required while exporting. Importing is probably suboptimal if we would unpack the file during each scan to read the manifest.

Reply:
> I personally find it a little silly to zip files and then use a custom extension (looks like a bad attempt at hiding something)

It's less about hiding something and more about saying "this is a mixxx-mapping" so that we could add an "Import Mapping' button to the controller preferences that allows picking a file from the downloads directory and automatically copies it to the right location without listing every single non-mixxx-related zipfile in the directory.

> Importing is probably suboptimal if we would unpack the file during each scan to read the manifest.

I suppose the performance impact is neglible. As an alternative, mixxx could provide an "export button" that zips the file w/o the manifest and includes the manifest text in the zip comment if we're worried about that.

Reply:
yeah, all pretty fair suggestions.

Reply:
> It's less about hiding something and more about saying "this is a mixxx-mapping"

I guess we could do something similar than for NI stem format and add a prefix to the extension (e.g `.mixxx.zip` or `.mixxx.tar.gz`). This would also us to support a wide range of compression format.

> Importing is probably suboptimal if we would unpack the file during each scan to read the manifest.

Probably getting too much into implementation details, but we could inspire from tools that uses a similar mechanism. Thinking of tools like Terraform or Helm that also have many "third party" extension compressed in an archive, you can expand them into a "cache" folder on import and maintain a checksum comparison on the source archive. As long as that checksum doesn't change, you can safely use the manifest file from the filesystem. This also greatly simplify modding as users can change the expanded module on the fly, without having to use a different mechanism (some kind of "draft" folder) or need to repack an archive.  

TL;DR; I don't think this is a problem.

Reply:
> This would also us to support a wide range of compression format.

This would really be a non-goal for me. The we need all these compression formats forever, and I think there is no good reason for it. Just agree on a single format (or rather two: packed for easy exchange and unpacked for easy development).

Reply:
Many file formats (like OpenDocument for Office application) use JAR files as container (just with a different file extension). The JAR file container defines the zipping and the manifest file format. And also handle topics like fast reading the manifest without unziping the whole files.

But I don't think that JAR files are good for the developer experience of a mapping developer. I suggest to use just sub-directories per mapping.